src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs

```diff
@@ -800,7 +800,15 @@ public static unsafe T[] AllocateUninitializedArray<T>(int length, bool pinned =
             if (pinned)
                 flags |= GC_ALLOC_FLAGS.GC_ALLOC_PINNED_OBJECT_HEAP;
 
-            return Unsafe.As<T[]>(AllocateNewArray(RuntimeTypeHandle.ToIntPtr(typeof(T[]).TypeHandle), length, flags));
+            return AllocateNewArrayWorker<T>(length, flags);
```

Comment:
Alternatively, JIT could skip the inline expansion of the pinvoke transition and fallback to dynamic stub for cold callsites driven by PGO?

Reply:
I think I've captured it back with my latest, unpushed PR.

Reply:
> fallback to dynamic stub for cold callsites driven by PGO

In this specific case, the cold path is likely going to be semi-worm. I doubt that PGO can solve this specific case well.

The general idea of taking PGO data into account for PInvoke inlining sounds good if we are not doing it already.